plugins/HostLinux.cpp

```diff
@@ -32,112 +32,127 @@ bool HostLinux::peek(const procptr_t address, void *dst, const size_t size) cons
 	return (ret != -1 && static_cast< size_t >(ret) == in.iov_len);
 }
 
-procptr_t HostLinux::module(const std::string &module) const {
-	std::stringstream ss;
-	ss << "/proc/";
-	ss << m_pid;
-	ss << "/maps";
-	const auto maps = readFile(ss.str());
+Modules HostLinux::modules() const {
+	std::ostringstream path;
+	path << "/proc/";
+	path << m_pid;
+	path << "/maps";
+	const auto maps = readFile(path.str());
 
 	if (maps.size() == 0) {
-		return 0;
+		return {};
 	}
 
-	std::stringstream ssPath(maps);
-	while (ssPath.good()) {
-		std::string baseaddr;
-
-		int ch;
-		while (1) {
-			ch = ssPath.get();
-			if (ch == '-') {
-				break;
-			} else if (ch == EOF) {
-				return 0;
+	Modules modules;
+	std::string name;
+	std::stringstream stream(maps);
+
+	while (stream.good()) {
+		std::string buf;
+
+		for (auto ch = stream.get(); ch != '-'; ch = stream.get()) {
+			if (stream.eof()) {
+				return modules;
 			}
-			baseaddr.push_back(static_cast< char >(ch));
+
+			buf.push_back(ch);
 		}
 
-		// seek to perms
-		do {
-			ch = ssPath.get();
-			if (ch == EOF) {
-				return 0;
+		MemoryRegion region;
+
+		region.address = std::stoull(buf, nullptr, 16);
+		buf.clear();
+
+		for (auto ch = stream.get(); ch != ' '; ch = stream.get()) {
+			if (stream.eof()) {
+				return modules;
 			}
-		} while (ch != ' ');
 
-		// seek to offset
-		do {
-			ch = ssPath.get();
-			if (ch == EOF) {
-				return 0;
+			buf.push_back(ch);
+		};
+
+		region.size = std::stoull(buf, nullptr, 16) - region.address;
+		buf.clear();
+
+		for (auto ch = stream.get(); ch != ' '; ch = stream.get()) {
+			if (stream.eof()) {
+				return modules;
 			}
```

Comment:
I think you can actually put the reading into its own function `readUntil` or something like that. That'd save some code duplication in your implementation :eyes:

Reply:
I agree.